ecdsa/src/ecdsa.rs

```diff
@@ -97,42 +101,54 @@ impl<E: CurveAffine, N: FieldExt, const NUMBER_OF_LIMBS: usize, const BIT_LEN_LI
         sig: &AssignedEcdsaSig<E::Scalar, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>,
         pk: &AssignedPublicKey<E::Base, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>,
         msg_hash: &AssignedInteger<E::Scalar, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>,
+        offsets: &mut HashMap<String, usize>,
     ) -> Result<(), Error> {
         let ecc_chip = self.ecc_chip();
         let scalar_chip = ecc_chip.scalar_field_chip();
         let base_chip = ecc_chip.base_field_chip();
 
+        offsets.insert("Started at".to_string(), ctx.offset());
+
         // 1. check 0 < r, s < n
 
         // since `assert_not_zero` already includes a in-field check, we can just
         // call `assert_not_zero`
+        offsets.insert("1. check 0 < r, s < n".to_string(), ctx.offset());
         scalar_chip.assert_not_zero(ctx, &sig.r)?;
         scalar_chip.assert_not_zero(ctx, &sig.s)?;
 
         // 2. w = s^(-1) (mod n)
+        offsets.insert("2. w = s^(-1) (mod n)".to_string(), ctx.offset());
         let (s_inv, _) = scalar_chip.invert(ctx, &sig.s)?;
 
         // 3. u1 = m' * w (mod n)
+        offsets.insert("3. u1 = m' * w (mod n)".to_string(), ctx.offset());
         let u1 = scalar_chip.mul(ctx, msg_hash, &s_inv)?;
 
         // 4. u2 = r * w (mod n)
+        offsets.insert("4. u2 = r * w (mod n)".to_string(), ctx.offset());
         let u2 = scalar_chip.mul(ctx, &sig.r, &s_inv)?;
 
         // 5. compute Q = u1*G + u2*pk
+        offsets.insert("5. compute Q = u1*G + u2*pk".to_string(), ctx.offset());
         let e_gen = ecc_chip.assign_point(ctx, Value::known(E::generator()))?;
         let g1 = ecc_chip.mul(ctx, &e_gen, &u1, 2)?;
         let g2 = ecc_chip.mul(ctx, &pk.point, &u2, 2)?;
         let q = ecc_chip.add(ctx, &g1, &g2)?;
 
         // 6. reduce q_x in E::ScalarExt
         // assuming E::Base/E::ScalarExt have the same number of limbs
+        offsets.insert("6. reduce q_x in E::ScalarExt".to_string(), ctx.offset());
         let q_x = q.x();
         let q_x_reduced_in_q = base_chip.reduce(ctx, q_x)?;
         let q_x_reduced_in_r = scalar_chip.reduce_external(ctx, &q_x_reduced_in_q)?;
 
         // 7. check if Q.x == r (mod n)
+        offsets.insert("7. check if Q.x == r (mod n)".to_string(), ctx.offset());
         scalar_chip.assert_strict_equal(ctx, &q_x_reduced_in_r, &sig.r)?;
```

Comment:
If one of these constraints fail with invalid data, you have to change them so they don't fail, while also doing in-circuit logic that the data/calculation is invalid. So if the circuit does something like `assert(x == 0)` (x needs to be `0`) you have to change it to `is_x_zero := x == 0` so x can be any value, but we can then use `is_x_zero` to mark some `is_signature_valid` value as false (all checks in circuit using constraints).